h2o-core/src/main/java/water/network/SSLSocketChannel.java

```diff
@@ -0,0 +1,358 @@
+package water.network;
+
+import water.H2O;
+import water.util.Log;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import java.io.IOException;
+import java.net.SocketException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ByteChannel;
+import java.nio.channels.SocketChannel;
+
+/**
+ * This class is based on:
+ * <a href="https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html">Oracle's JSSE guide.</a>
+ * <a href="https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/samples/sslengine/SSLEngineSimpleDemo.java">Oracle's SSLEngine demo.</a>
+ *
+ * It's a simple wrapper around SocketChannels which enables SSL/TLS
+ * communication using {@link javax.net.ssl.SSLEngine}.
+ */
+class SSLSocketChannel implements ByteChannel {
+
+    // Empty buffer for handshakes
+    private static final ByteBuffer EMPTY_BUFFER = ByteBuffer.allocate(0);
+
+    // Buffer holding encrypted outgoing data
+    private ByteBuffer netInBuffer;
+    // Buffer holding encrypted incoming data
+    private ByteBuffer netOutBuffer;
+
+    // Buffer holding decrypted incoming data
+    private ByteBuffer peerAppData;
+
+    private SocketChannel channel = null;
+    private SSLEngine sslEngine = null;
+
+    private boolean closing = false;
+    private boolean closed = false;
+
+    private boolean handshakeComplete = false;
+
+    SSLSocketChannel(SocketChannel channel, SSLEngine sslEngine) throws IOException {
+        this.channel = channel;
+        this.sslEngine = sslEngine;
+
+        sslEngine.setEnableSessionCreation(true);
+        SSLSession session = sslEngine.getSession();
+        prepareBuffers(session);
+
+        handshake();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return channel.isOpen();
+    }
+
+    @Override
+    public void close() throws IOException {
+        closing = true;
+        sslEngine.closeOutbound();
+        sslEngine.getSession().invalidate();
+        netOutBuffer.clear();
+        channel.close();
+        closed = true;
+    }
+
+    private void prepareBuffers(SSLSession session) throws SocketException {
+        int appBufferSize = session.getApplicationBufferSize();
+        // Less is not more. More is more. Bigger than the app buffer size so successful unwraps() don't cause BUFFER_OVERFLOW
+        peerAppData = ByteBuffer.allocate(appBufferSize + 64);
```

Comment:
@mmalohlava not really, saw other libraries use this value and from my testing it was big enough to prevent any overflows